### PR TITLE
Fix incompatibility with the attr encrypted gem

### PR DESCRIPTION
**Pull request by we138**
Hi! I found this [issue](https://github.com/thoughtbot/clearance/issues/685), and I try to fix it in this way:
private method `encrypt` have a one line and is used once in the setter-method `password=()`. I suggest to remove method `encrypt` and used `::BCrypt::Password.create(password, cost: cost)` straight in to `password=()`.

### CODE REVIEW

**Comment by magikid** [comment]
I don't think this is a good solution.  One of the reasons that the `encrypt` method is there is so that you can swap BCrypt out for a different strategy if you want.

**Comment by we138** [comment]
@magikid Hi Chris, you absolutely right, I didn't noticed it. What do you think, how can fix it issue?

**Comment by theharq** [comment]
Hello! and thank you for doing this first attempt to fix this method name collision. 💚 

Maybe I'm missing something but the method `encrypt` is not part of the public api of the module.

As the documentation says:

>   # You can supply your own password strategy by implementing a module that
>   # responds to the proper interface methods. Once this module is configured as
>   # your password strategy, Clearance will mix it into your Clearance User
>   # class. Thus, your module can access any methods or attributes on User.
>   #
>   # Password strategies need to respond to `authenticated?(password)` and
>   # `password=(new_password)`. For an example of how to implement these methods,
>   # see {Clearance::PasswordStrategies::BCrypt}.
> https://github.com/theharq/clearance/blob/master/lib/clearance/password_strategies.rb

So in our custom module as long we respond to `authenticated?(password)` and `password=(new_password)` and set this module in the configuration block.

```
Clearance.configure do |config|
  config.password_strategy = Clearance::PasswordStrategies::Custom
end
```

We could swap the default `PasswordStrategies::BCrypt`

This change may not be necessary for the other modules because all clearance-provided password strategies other than BCrypt have been deprecated.

Maybe we could update the [fake_model_with_password_strategy.rb](https://github.com/thoughtbot/clearance/blob/master/spec/support/fake_model_with_password_strategy.rb) to include the `encrypt` method so we make sure all the tests pass.. but I don't think the maintainers will want this.

Maybe there is another option to add a failing test for this edge case.

Thank you again, for taking the time to propose a solution 👍

**Comment by magikid** [comment]
That's a good point @theharq.  I guess I should have read the comments a little better.

Because it's private api, I think a simple fix would be to just change the `encrypt` method name to something like `bcrypt_encrypt` (and changing it in the few spots that call `encrypt`).  Do you think that would be a decent solution?

**Comment by theharq** [comment]
@magikid yeah that will be a good solution having a *strategy*`_encrypt` or `encrypt_password` method will do the trick, but still is up to the project maintainers and their plans for clearance 2.0

**Comment by we138** [comment]
@theharq @mackuba hi guys, I have been renamed `encrypt` methods in whole strategies

**Comment by mackuba** [comment]
Clearance? Hmm...



;)

**Comment by tadp** [comment]
Thanks for tackling this.

I noticed this PR has no assignees but is in need of review. Maybe @derekprior can assign someone?

**Comment by ACPK** [comment]
@magikid @tadp  - What is your recommended solution for this as of right now?

**Comment by tadp** [comment]
@ACPK I like the renaming that @theharq suggested and that @we138 implemented in the latest commit.

`_` is [commonly used](https://www.quora.com/Why-using-underscore-to-represent-private-variable-method) in other languages to imply a private method and it is unlikely to be exposed by other gems

One thing I do see that this PR needs is a test to prevent this sort of thing from happening again by someone accidentally renaming the funny looking method back.

Thanks @we138 for doing this! WDYT about:
- [ ] adding this drop-in test to your PR that tests the method naming. It can be placed in `rspec spec/password_strategies/bcrypt_spec.rb:72`. I tried it out on your branch and it works:

```
  describe "#_encrypt" do
    it "reserves #encrypt to prevent name conflicts with other gems" do
      model_instance = fake_model_with_bcrypt_strategy

      expect(model_instance.respond_to?(:encrypt, true)).to eq(false)
    end

    it "responds to #_encrypt" do
      model_instance = fake_model_with_bcrypt_strategy

      expect(model_instance.respond_to?(:_encrypt, true)).to eq(true)
    end

    it "ensures that #_encrypt is a private method as its name implies" do
      model_instance = fake_model_with_bcrypt_strategy

      expect(model_instance.respond_to?(:_encrypt)).to eq(false)
    end
  end
```
- [ ] [Multiline commit message](http://chris.beams.io/posts/git-commit/) that explains the issue squashed into a single commit. This would make this really easy to revert or track down later.

I think with those two things this would be good for @derekprior to review.

**Comment by magikid** [comment]
@acpk I believe the solution is to rename the `encrypt` method like this PR does.

**Comment by we138** [comment]
@tadp yo Tad, your task is done!

**Comment by tadp** [comment]
@we138 I think this is great. This PR is following the [contributing guidelines](https://github.com/thoughtbot/clearance/blob/master/CONTRIBUTING.md) well. However, I did not check off as completed "multiline commit message" in my previous comment because the commit message is still one line.

Maybe on the second line of the commit message it can say the WHY and HOW of the change?  It fixes incompatibility how?  It may be easy for us to see how it does now, but if someone does a git blame on the project months from now there is not much context on what is going on.

Here is the [thoughtbot styleguide](https://github.com/thoughtbot/guides/tree/master/style/git) on this. A short explaination should be fine.

**Comment by derekprior** [comment]
Just wanted to note that I'm not ignoring this, but that I have had zero time for open source work of late. I will get to this, but it likely will not be until later this month. Thanks for understanding.

File: lib/clearance/password_strategies/blowfish.rb [hunk 1]
```diff
@@ -15,7 +15,7 @@ module Blowfish
       #   gem
       def authenticated?(password)
         warn "#{Kernel.caller.first}: #{DEPRECATION_MESSAGE}"
-        encrypted_password == encrypt(password)
+        encrypted_password == _encrypt(password)
```
Comment (derekprior) [inline lib/clearance/password_strategies/blowfish.rb]: The blowfish password strategy is deprecated. Please do not make any changes to this file. Thanks!

File: lib/clearance/password_strategies/sha1.rb [hunk 2]
```diff
@@ -16,7 +16,7 @@ module SHA1
       #   gem
       def authenticated?(password)
         warn "#{Kernel.caller.first}: #{DEPRECATION_MESSAGE}"
-        encrypted_password == encrypt(password)
+        encrypted_password == _encrypt(password)
```
Comment (derekprior) [inline lib/clearance/password_strategies/sha1.rb]: The blowfish password strategy is deprecated. Please do not make any changes to this file. Thanks!

File: spec/password_strategies/bcrypt_spec.rb [hunk 3]
```diff
@@ -69,6 +69,26 @@ def encrypted_password
     end
   end
 
+  describe "#_encrypt" do
```
Comment (derekprior) [inline spec/password_strategies/bcrypt_spec.rb]: As long as the existing tests pass, I'm happy to skip specific unit tests for the private method.

**Comment by derekprior** [comment]
This is a rather unfortunate side effect of password strategies being implemented as mixins to the User class rather than their own classes. 

Renaming `encrypt` solves this particular problem, but really we shouldn't be mixing in the methods at all. They are not useful to the user's `User` class. I opened a new PR where inline those methods: https://github.com/thoughtbot/clearance/pull/693

**Comment by derekprior** [comment]
I went with the approach in #693 instead, but this is now fixed. Thanks so much to everyone involved in chasing this down.

I will try to release soon but I have spotty/unreliable wifi for the time being.

**Comment by tadp** [comment]
Yeah sounds great.  Thanks @derekprior!! 💯 

And @we138 for the first pass, @theharq for bringing it up and the eyes of @magikid, @mackuba @ACPK.

**Comment by derekprior** [comment]
1.14.2 is out with this fix. Thanks.
